source/basic_skills.rst

```diff
@@ -1304,32 +1304,595 @@ If you managed to solve this, congratulations!
 You are well on your way to becoming a master formalizer.
 
 
-.. .. _using_more_theorems_and_lemmas:
+.. _more_examples:
 
-.. Using More Theorems and Lemmas
-.. ------------------------------
+More Examples
+-------------
 
-.. [I got tired of writing, so I decided to give the reader a break
-.. and start a new section.
-.. But I have more fun examples and exercises:
-.. things with ``min`` and ``max``,
-.. things with ``dvd`` and ``gcd``,
-.. and an exercise with a variant of the triangle inequality.]
+The ``min`` function on the real numbers is uniquely characterized
+by the following three facts:
 
-.. .. _proving_facts_about_algebraic_structures:
+.. code-block:: lean
+
+    import data.real.basic
+
+    variables a b c d : ℝ
+
+    -- BEGIN
+    #check (min_le_left a b : min a b ≤ a)
+    #check (min_le_right a b : min a b ≤ b)
+    #check (le_min : c ≤ a → c ≤ b → c ≤ min a b)
+    -- END
+
+Can you guess the names of the theorems that characterize
+``max`` in a similar way?
+
+Using the theorem ``le_antisymm``, we can show that two
+real numbers are equal if each is less than or equal to the other.
+Using this and the facts above,
+we can show that ``min`` is commutative:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    variables a b : ℝ
+
+    -- BEGIN
+    example : min a b = min b a :=
+    begin
+      apply le_antisymm,
+      { show min a b ≤ min b a,
+        apply le_min,
+        { apply min_le_right },
+        apply min_le_left },
+      { show min b a ≤ min a b,
+        apply le_min,
+        { apply min_le_right },
+        apply min_le_left }
+    end
+    -- END
+
+Here we have used curly brackets to separate proofs of
+different goals.
+Our usage is inconsistent:
+at the outer level,
+we use curly brackets and indentation for both goals,
+whereas for the nested proofs,
+we use curly brackets only until a single goal remains.
```

Comment:
Actually I never do that. I use curly brackets also around the last remaining goal.

Reply:
But I do. The problem is that if you have a long proof and lots of the theorems have side conditions, with curly braces you creep across the page for no good reason. With this convention, your proofs stay linear:
```
  apply foo,
  { side condition },
  { side condition }, 
  { side condition },
  back to the main proof, 
  ...
```
I learned this style from Georges Gonthier; it is a convention in Mathematical Components.